resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala

```diff
@@ -192,58 +225,94 @@ private[spark] class KubernetesClusterSchedulerBackend(
+            .addToLimits("memory", executorMemoryLimitQuantity)
+            .addToRequests("cpu", executorCpuQuantity)
+            .addToLimits("cpu", executorCpuQuantity)
+          .endResources()
+          .withEnv(requiredEnv.asJava)
+          .withPorts(requiredPorts.asJava)
+        .endContainer()
+      .endSpec()
+
+    var resolvedPodBuilder = shuffleServiceConfig.map { config =>
+      config.shuffleDirs.foldLeft(basePodBuilder) { (builder, dir) =>
+        builder.editSpec()
+          .addNewVolume()
+            .withName(FilenameUtils.getBaseName(dir))
+            .withNewHostPath().withPath(dir)
+            .endHostPath()
+          .endVolume()
+          .editFirstContainer()
+            .addNewVolumeMount()
+            .withName(FilenameUtils.getBaseName(dir))
+            .withMountPath(dir)
+            .endVolumeMount()
             .endContainer()
           .endSpec()
-        .done())
+      }
+    }.getOrElse(basePodBuilder)
+
+    try {
+      (executorId, kubernetesClient.pods().create(resolvedPodBuilder.build()))
     } catch {
       case throwable: Throwable =>
         logError("Failed to allocate executor pod.", throwable)
         throw throwable
     }
   }
 
-  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] = Future[Boolean] {
-    EXECUTOR_MODIFICATION_LOCK.synchronized {
-      if (requestedTotal > totalExpectedExecutors.get) {
-        logInfo(s"Requesting ${requestedTotal - totalExpectedExecutors.get}"
-          + s" additional executors, expecting total $requestedTotal and currently" +
-          s" expected ${totalExpectedExecutors.get}")
-        for (i <- 0 until (requestedTotal - totalExpectedExecutors.get)) {
-          runningExecutorPods += allocateNewExecutorPod()
+  override def doRequestTotalExecutors(requestedTotal: Int): Future[Boolean] =
+    Future[Boolean] {
+      EXECUTOR_MODIFICATION_LOCK.synchronized {
+        if (requestedTotal > totalExpectedExecutors.get) {
+          logInfo(s"Requesting ${requestedTotal - totalExpectedExecutors.get}"
+            + s" additional executors, expecting total $requestedTotal and currently" +
+            s" expected ${totalExpectedExecutors.get}")
+          for (i <- 0 until (requestedTotal - totalExpectedExecutors.get)) {
+            runningExecutorPods += allocateNewExecutorPod()
+          }
+        }
+        totalExpectedExecutors.set(requestedTotal)
+      }
+
+      var allPodsReady = true
+      for (executorPod <- runningExecutorPods.values) {
+        try {
+          val pod = kubernetesClient.pods().inNamespace(kubernetesNamespace)
+            .withName(executorPod.getMetadata().getName()).get()
+
+          allPodsReady &&= Readiness.isPodReady(pod)
+        } catch {
+          case throwable: Throwable =>
+            logWarning(s"Executor no longer exists.", throwable)
+            allPodsReady = false
         }
       }
-      totalExpectedExecutors.set(requestedTotal)
+
+      allPodsReady
```

Comment:
There is one difference here I think. Kubernetes will not refuse to accept new pods even if there are no resources, and they'll just go pending. We need some mechanism in between to ensure we do not request more executors than can be accommodated at the current time.